.github/workflows/deploy-functions.yml

```
@@ -0,0 +1,132 @@
+name: Deploy Python Firebase Functions
+
+on:
+  push:
+    branches:
+      - add-deploy-workflow
+  workflow_dispatch:
+    inputs:
+      environment:
+        description: 'Environment to deploy to'
+        required: true
+        default: 'staging'
+        type: choice
+        options:
+          - staging
+          - prod
+      confirm_prod:
+        description: 'Type "yes" to confirm production deployment'
+        required: false
+        type: string
+
+permissions:
+  contents: read
+
+concurrency:
+  group: deploy-${{ github.ref }}
+  cancel-in-progress: false
+
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: Validate production deployment
+        if: ${{ github.event.inputs.environment == 'prod' }}
+        run: |
+          if [[ "${{ github.event.inputs.confirm_prod }}" != "yes" ]]; then
+            echo "::error::Production deployment requires explicit confirmation. Please type 'yes' in the confirm_prod field."
+            exit 1
+          fi
+
+      - name: Set up Node.js for Firebase CLI
+        uses: actions/setup-node@v3
+        with:
+          node-version: '20'
+
+      - name: Install Firebase CLI
+        run: |
+          npm install -g firebase-tools
+          firebase --version
+
+      - name: Install uv package manager
+        uses: astral-sh/setup-uv@v5
+        with:
+          version: "0.6.4"
+
+      # Generate requirements.txt for Firebase Functions
+      - name: Generate requirements.txt
+        working-directory: functions
+        run: uv pip compile pyproject.toml -o requirements.txt
+
+      - name: Set up Python 3.11
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.11"
+
+      - name: Install dependencies with pip
+        working-directory: functions
+        run: |
+          python -m venv venv
+          source venv/bin/activate
+          pip install --upgrade pip
+          pip install -r requirements.txt
```

We are using a shiny new tool `uv`, which supports `pip` for antiquated reasons. We can simplify this by using the new tool.

```suggestion
      - name: Set up Python
         run: uv python install
```
This _should_ do all the work instead.

---

Ahh, I think my suggestion is missing something. `uv` will create a `.venv`, but firebase requires a `venv`. To fix this, we need to run `ln -s .venv venv`.

---

I tried deploying the Function with the symlink (locally and on Firebase), but got:
`FirebaseError: Could not detect runtime for functions`

I found a way to install dependencies on the runner using `uv` exclusively. Firebase expects a virtual environment named `venv` with `pip` installed within it and a `requirements.txt`. I also needed to update the GitHub Actions workflow to use the correct environment for `uv sync` with the `--active` flag.